src/dataframe.rs

```diff
@@ -86,9 +87,8 @@ pub struct ReadCsvOptions {
     pub eol_char: String,
 }
 
-#[napi(catch_unwind)]
-pub fn read_csv(
-    path_or_buffer: Either<String, Buffer>,
+fn mmap_reader_to_df<'a>(
+    csv: impl FnOnce() -> Box<dyn MmapBytesReader> + 'a,
```

Comment:
why can't we just have this take in `impl MmapBytesReader` instead of boxing it and wrapping it in a closure?

```rs
fn mmap_reader_to_df<'a>(csv: impl MmapBytesReader<'a>)
```

Reply:
I am getting bunch of errors when I do it.
```
error[E0277]: the trait bound `{closure@src/dataframe.rs:177:13: 177:15}: MmapBytesReader` is not satisfied
   --> src/dataframe.rs:177:13
    |
176 |           Either::B(buffer) => mmap_reader_to_df(
    |                                ----------------- required by a bound introduced by this call
177 | /             || {
178 | |                 let bytes = unsafe { std::mem::transmute::<&'_ [u8], &'static [u8]>(&buffer) };
179 | |                 Cursor::new(bytes.to_owned())
180 | |             },
    | |_____________^ the trait `MmapBytesReader` is not implemented for closure `{closure@src/dataframe.rs:177:13: 177:15}`
    |
note: required by a bound in `mmap_reader_to_df`
```
The boxing syntax from py-polars [implementation](https://github.com/pola-rs/polars/blob/a4d30b4a500394fcb421ec39919cdec97fe8e2e1/py-polars/src/file.rs#L221).

Reply:
the py implementation is a bit different here, we only allow `Either<String, Buffer>` So we should be able to accept the following signature

```rs
fn mmap_reader_to_df<'a>(
    csv: impl MmapBytesReader + 'a,
    options: ReadCsvOptions,
) -> napi::Result<JsDataFrame>;
```

We also shouldn't transmute here. There is no need. `Buffer` implements `AsRef` for `&[u8]`

```rs
Either::B(buffer) => {
    let cursor = Cursor::new(buffer.as_ref());
    mmap_reader_to_df(cursor, options)
},
```